### PR TITLE
Exit if STDIN is closed #1367

### LINKED ISSUE #1367: Ranger enters busy-loop if STDIN file descriptor is closed
I hope this will solve an old heisenbug. It doesn't seem to always reproduce naturally, but can be provoked using gdb. 

Potentially related, unresolved bug reports: 
- https://github.com/ranger/ranger/issues/87
- https://github.com/ranger/ranger/issues/525

#### Runtime Environment
- Operating system and version: Ubuntu 18.04.1 LTS
- Terminal emulator and version: rxvt-unicode (urxvt) v9.22 - released: 2016-01-23
- ranger version: ranger-master v1.9.2-53-ge56c74bb
- Python version: 2.7.15rc1 (default, Apr 15 2018, 21:51:34) [GCC 7.3.0]
- Locale: None.None
- Shell: GNU bash, version 4.4.19(1)-release (x86_64-pc-linux-gnu)

#### Current Behavior

Ranger enters a busy-loop if STDIN is closed. 

#### Expected Behavior

Ranger exits if STDIN is closed. 

#### Background

Using ranger extensively, I occasionally notice that a ranger process is using 100% of the CPU. 

It is not uncommon for me to start ranger, navigate to a directory, spawn a shell to run a command and restart ranger again. 
I also usually close terminal windows using the window manager, rather than exiting each subprocess running in that window. 

#### Analysis

To the best of my understanding, this is how the bug manifests. 

When a shell exits (e.g. when closing a terminal), it sends a SIGHUP signal to its child processes. 
If the child process does not explicitly handle this signal (which ranger doesn't do), the default behavior is to terminate the process. 

When ranger spawns a shell which restarts ranger, the SIGHUP reaches the first ranger process, but not (always?) the second process (due to the fork-exec?). The terminal and the first ranger process exit, and the remaining ranger process continues to run, but becomes orphaned and adopted by init (pid 1). 

Ranger runs a 'while True' loop which repeatedly wait for user input, which happens here: https://github.com/ranger/ranger/blob/e56c74bb30ad2f363280e4d1a14bf2c20dff5b5f/ranger/gui/ui.py#L239

	key = self.win.getch()

https://docs.python.org/2/library/curses.html#curses.window.getch

If there is no input, getch() waits for some time, then returns -1. However, because the STDIN file descriptor was closed by the earlier exited processes, getch() presumably throws an error and returns -1 immediately, which causes ranger to enter a continuous busy-loop. 

#### Possible Solutions

If getch() returns -1 and os.isatty(0) returns False, ranger should exit. 

##### Natural way (not guaranteed to work)

In a terminal, start ranger: 

  > ranger

Press 'S' to spawn a shell
Start ranger again:

  > ranger

(Maybe press down a couple of times)
Close the terminal (e.g. Alt-F4)

The former ranger process exits, but the latter enters a busy-loop. 

##### gdb way

Force-close the STDIN file descriptor of a running ranger process

Start ranger
Attach to process with gdb: 

  > gdb -p PID_OF_RANGER

Close fd 0:

  > (gdb) p close(0)

Detach to let ranger keep running:

  > (gdb) detach

Ranger enters busy-loop.

### PR DESCRIPTION
**Pull request by igoyak**
See https://github.com/ranger/ranger/issues/1367

I'm not sure if this is the preferred way to exit ranger, suggestions welcome

### CODE REVIEW

File: ranger/gui/ui.py [hunk 1]
```diff
@@ -272,6 +272,9 @@ def handle_input(self):
                 else:
                     if not self.fm.input_is_blocked():
                         self.handle_key(key)
+            elif key == -1 and not os.isatty(0):
```
Comment (toonn) [inline ranger/gui/ui.py]: Could you use `sys.stdin.fileno()` rather than the hard coded `0`?

File: ranger/gui/ui.py [hunk 2]
```diff
@@ -272,6 +272,9 @@ def handle_input(self):
...
+                # STDIN has been closed
+                sys.exit(1)
```
Comment (toonn) [inline ranger/gui/ui.py]: Proper way to exit is `self.fm.exit()`

**Comment by toonn** [comment]
I'm pretty sure the other issues are the same problem. Feel free to include `Fixes #1367\nFixes #87...` in one of your commits.

**Comment by igoyak** [comment]
Thanks for the feedback! I will check it out in a couple of days.

File: ranger/gui/ui.py (hunk 2 above)
Comment (igoyak) [reply ranger/gui/ui.py]: Thanks, fixed.

File: ranger/gui/ui.py (hunk 1 above)
Comment (igoyak) [reply ranger/gui/ui.py]: Good idea, fixed.

**Comment by toonn** [comment]
Hmm, I actually can't reproduce the problem with the master version of ranger. @SuperShinyEyes, could you test this fix?

**Comment by igoyak** [comment]
About propagating the SIGHUP: I've done some reading, but I'm not entirely sure how to do it, why it isn't happening currently or whether it's actually a good idea. It is related to the "job control" of shells and "process groups", where the "process leader" is responsible for forwarding signals (or rather translate a SIGHUP into SIGINTs for the children?). I could try to dig deeper, but maybe not in the foreseeable future. 

I do believe the behavior proposed in this PR is correct, though. Reasoning: ranger is a user-facing application. If the user cannot supply input, ranger has no reason to keep running.

**Comment by igoyak** [comment]
> Hmm, I actually can't reproduce the problem with the master version of ranger. @SuperShinyEyes, could you test this fix?

Did you try the method using gdb? That is the only one that seems reliable. 
I rechecked both methods and can still produce the bug in commit 1ca1255a3cf26eb646fca75b5388b2c645fdf466

**Comment by SuperShinyEyes** [comment]
@toonn @igoyak Is there a still need for me to test this fix? 

Sorry for the late reply. Last week was BUSY. I'd be more than happy to help fixing this. :)

**Comment by SuperShinyEyes** [comment]
@toonn Got it. Will try tonight.

**Comment by igoyak** [comment]
@toonn I have finally had more time to look deeper into this issue. 
I have updated the source branch of this PR, just rebased to use a modern commit as a parent. 
I have also written some more conclusions and a cross-platform way to replicate the issue: https://github.com/ranger/ranger/issues/1367#issuecomment-521773193

This issue still affects me (and others, it seems), and I think it would be a good idea to merge to master whenever appropriate.

**Comment by toonn** [comment]
Will look into this tomorrow. Thanks for your effort!

@SuperShinyEyes, did you find the time to test this?

**Comment by toonn** [comment]
Hmm, finally found time to look into this. Still can't reproduce by closing a terminal.
Apparently gdb won't run on macOS due to some code signing issues, at least without much effort. However, I was able to reproduce this with gdb on linux, had to guess at the type to cast the `close` call to though.

Now I can finally merge this 🎉 : )

**Pull request merged**